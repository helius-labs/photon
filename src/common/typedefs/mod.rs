pub mod hash;
pub mod serializable_pubkey;
