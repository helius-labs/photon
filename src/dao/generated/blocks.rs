//! `SeaORM` Entity. Generated by sea-orm-codegen 0.10.6

use sea_orm::entity::prelude::*;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "blocks")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub slot: i64,
    pub parent_slot: i64,
    pub parent_blockhash: Vec<u8>,
    pub blockhash: Vec<u8>,
    pub block_height: i64,
    pub block_time: i64,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::transactions::Entity")]
    Transactions,
}

impl Related<super::transactions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Transactions.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
