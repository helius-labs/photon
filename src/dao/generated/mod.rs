//! `SeaORM` Entity. Generated by sea-orm-codegen 0.10.6

pub mod prelude;

pub mod account_transactions;
pub mod accounts;
pub mod blocks;
pub mod owner_balances;
pub mod state_trees;
pub mod token_accounts;
pub mod token_owner_balances;
pub mod transactions;
