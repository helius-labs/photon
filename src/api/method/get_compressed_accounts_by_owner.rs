use crate::dao::generated::accounts;
use sea_orm::{ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter, QueryOrder, QuerySelect};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use super::{
    super::error::PhotonApiError,
    utils::{Context, Limit, PAGE_LIMIT},
};
use crate::common::typedefs::{hash::Hash, serializable_pubkey::SerializablePubkey};

use super::utils::{parse_account_model, Account};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct Options {
    pub cursor: Option<Hash>,
    pub limit: Option<Limit>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct GetCompressedAccountsByOwnerRequest {
    pub owner: SerializablePubkey,
    pub cursor: Option<Hash>,
    pub limit: Option<Limit>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct PaginatedAccountList {
    pub items: Vec<Account>,
    pub cursor: Option<Hash>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct GetCompressedAccountsByOwnerResponse {
    pub context: Context,
    pub value: PaginatedAccountList,
}

pub async fn get_compressed_accounts_by_owner(
    conn: &DatabaseConnection,
    request: GetCompressedAccountsByOwnerRequest,
) -> Result<GetCompressedAccountsByOwnerResponse, PhotonApiError> {
    let context = Context::extract(conn).await?;
    let GetCompressedAccountsByOwnerRequest {
        owner,
        cursor,
        limit,
    } = request;

    let mut filter = accounts::Column::Owner
        .eq::<Vec<u8>>(owner.into())
        .and(accounts::Column::Spent.eq(false));

    if let Some(cursor) = cursor {
        filter = filter.and(accounts::Column::Hash.gt::<Vec<u8>>(cursor.into()));
    }
    let mut query_limit = PAGE_LIMIT;
    if let Some(limit) = limit {
        query_limit = limit.value();
    }

    let result = accounts::Entity::find()
        .order_by(accounts::Column::Hash, sea_orm::Order::Asc)
        .limit(query_limit)
        .filter(filter)
        .all(conn)
        .await?;

    let items = result
        .into_iter()
        .map(parse_account_model)
        .collect::<Result<Vec<Account>, PhotonApiError>>()?;

    let mut cursor = items.last().map(|u| u.hash.clone());
    if items.len() < query_limit as usize {
        cursor = None;
    }

    Ok(GetCompressedAccountsByOwnerResponse {
        context,
        value: PaginatedAccountList { items, cursor },
    })
}
