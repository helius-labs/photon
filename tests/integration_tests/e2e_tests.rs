use function_name::named;
use photon_indexer::api::method::get_compressed_accounts_by_owner::GetCompressedAccountsByOwnerRequest;
use photon_indexer::api::method::get_transaction_with_compression_info::get_transaction_helper;
use photon_indexer::common::typedefs::serializable_pubkey::SerializablePubkey;
use photon_indexer::ingester::index_block;
use solana_sdk::pubkey::Pubkey;

use crate::utils::*;
use insta::assert_json_snapshot;
use photon_indexer::api::method::utils::GetCompressedTokenAccountsByOwner;
use photon_indexer::common::typedefs::hash::Hash;
use photon_indexer::dao::generated::blocks;
use photon_indexer::ingester::typedefs::block_info::{BlockInfo, BlockMetadata};
use sea_orm::ColumnTrait;
use sea_orm::EntityTrait;
use sea_orm::QueryFilter;
use serial_test::serial;

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_e2e_mint_and_transfer(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    use std::str::FromStr;

    
    use photon_indexer::{
        api::method::{
            get_compression_signatures_for_token_owner::GetCompressionSignaturesForTokenOwnerRequest,
            get_multiple_compressed_account_proofs::HashList, utils::Limit,
        },
        common::typedefs::serializable_signature::SerializableSignature,
    };
    use solana_sdk::signature::Signature;

    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let payer_pubkey =
        SerializablePubkey::try_from("Fg5TtbkvcqLedmwFqqbWty9NRqMKKMqy9zNLZZRQXSDc").unwrap();
    let bob_pubkey =
        SerializablePubkey::try_from("4rANawp9ahHmvaVi3CCeJSam8JTA4DDY3htJKnhYxbHN").unwrap();
    let charles_pubkey =
        SerializablePubkey::try_from("GyXNDaUUBwzYp24vEzang6SyCzcA8L3csZkiALqF6KRs").unwrap();

    let mint_tx =
        "2DcgLz1KwPaMoiRAAmYAirx4213swLbaSM1YYuCkbgzngq61XSWaFCRGuFoMgqW5htsEZRKBm6uQQPAPCXUVL8eH";
    let transfer_tx =
        "DFE7shhwCWd5P14NfA1ELSpd96FaJhBkjSet2Y7t7CtM4oLhP17AYKtyMBjZWi4hgFKvqET8zcAQUDuAC6sARfk";
    let transfer_txn2: &str =
        "5FDV6xrPRyfEUP2Kg7mwZVj34LVEfq9gsrquyFkvmmdPiiEn3Bhd8sTJTe6UhszNHLgwWfV6j9X4rsceL9mNNyRZ";
    let transfer_txn3 =
        "2pGQ7MgWuhrKpeSAzs58CXttKDiws33X68hQv4tRm6NUH6JCsCBk5uteNcjt2fXdToYhC6s8b6SDzoXV9CLJTo94";

    let txs = [mint_tx, transfer_tx, transfer_txn2, transfer_txn3];

    // HACK: We index a block so that API methods can fetch the current slot.
    index_block(
        &setup.db_conn,
        &BlockInfo {
            metadata: BlockMetadata {
                slot: 0,
                ..Default::default()
            },
            ..Default::default()
        },
    )
    .await
    .unwrap();

    for tx in txs {
        index_transaction(&setup, tx).await;
    }
    for (person, pubkey) in [
        ("bob", bob_pubkey.clone()),
        ("charles", charles_pubkey.clone()),
        ("payer", payer_pubkey.clone()),
    ] {
        let accounts = setup
            .api
            .get_compressed_token_accounts_by_owner(GetCompressedTokenAccountsByOwner {
                owner: pubkey.clone(),
                ..Default::default()
            })
            .await
            .unwrap();
        assert_json_snapshot!(format!("{}-{}-accounts", name.clone(), person), accounts);
        let proofs = setup
            .api
            .get_multiple_compressed_account_proofs(HashList(
                accounts
                    .value
                    .items
                    .iter()
                    .map(|x| x.account.hash.clone())
                    .collect(),
            ))
            .await
            .unwrap();
        assert_json_snapshot!(format!("{}-{}-proofs", name.clone(), person), proofs);

        let mut cursor = None;
        let limit = Limit::new(1).unwrap();
        let mut signatures = Vec::new();
        loop {
            let res = setup
                .api
                .get_compression_signatures_for_token_owner(
                    GetCompressionSignaturesForTokenOwnerRequest {
                        owner: pubkey.clone(),
                        cursor,
                        limit: Some(limit.clone()),
                    },
                )
                .await
                .unwrap()
                .value;
            signatures.extend(res.items);
            cursor = res.cursor;
            if cursor.is_none() {
                break;
            }
        }
        assert_json_snapshot!(
            format!("{}-{}-token-signatures", name.clone(), person),
            signatures
        );

        let token_balances = setup
                .api
                .get_compressed_token_balances_by_owner(photon_indexer::api::method::get_compressed_token_balances_by_owner::GetCompressedTokenBalancesByOwnerRequest {
                    owner: pubkey.clone(),
                    ..Default::default()
                })
                .await
                .unwrap();

        assert_json_snapshot!(
            format!("{}-{}-token-balances", name.clone(), person),
            token_balances
        );
    }
    for (txn_name, txn_signature) in [("mint", mint_tx), ("transfer", transfer_tx)] {
        let txn = cached_fetch_transaction(&setup, txn_signature).await;
        let txn_signature = SerializableSignature(Signature::from_str(txn_signature).unwrap());
        // Test get transaction
        let parsed_transaction: photon_indexer::api::method::get_transaction_with_compression_info::GetTransactionResponse = get_transaction_helper(&setup.db_conn, txn_signature, txn).await.unwrap();
        assert_json_snapshot!(
            format!("{}-{}-transaction", name.clone(), txn_name),
            parsed_transaction
        );
    }
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_lamport_transfers(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    use photon_indexer::api::method::get_multiple_compressed_account_proofs::HashList;

    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Localnet,
            db_backend,
        },
    )
    .await;

    let compress_tx =
        "3ZmBETHr8tHPr7iKdgW4NjTzcUCUHtSBFQD2JwSniShBV9fix6xb6FxRyR1cMRW83WerRYU5b1uxDLM3GqomCkYx";
    let transfer_tx1 =
        "64tDNpAnQHZsuiH8UG8vXDzXbwGHdghfLYerPdaVdgkQ2kSTE65kDT7ajzzhd2ZDW8nVoKYztnjxCaV8bWJxNCXR";
    let transfer_tx2 =
        "yK3j6TkABz2SVBtvsVwJV5ZoLv1tJBrBR8Z8JsosHQGbeizt7yx8Qx1Nva1ytcKa7hAXsT7R6WyQUiCH43Cwtk5";

    let payer_pubkey = SerializablePubkey(
        Pubkey::try_from("8wSvtG3dvL7Nso8E3wx6L9SwTRiZuR8ojRX5D6mwx7hJ").unwrap(),
    );
    let receiver_pubkey = SerializablePubkey(
        Pubkey::try_from("G1zCvDhtSdV6m92q3tXBEE4d531fBx1vD7EUpeeQZrNp").unwrap(),
    );
    let txs = [compress_tx, transfer_tx1, transfer_tx2];
    let number_of_indexes = [1, 2];

    for individually in [true, false] {
        reset_tables(&setup.db_conn).await.unwrap();
        for number in number_of_indexes.iter() {
            for _ in 0..*number {
                // HACK: We index a block so that API methods can fetch the current slot.
                index_block(
                    &setup.db_conn,
                    &BlockInfo {
                        metadata: BlockMetadata {
                            slot: 0,
                            ..Default::default()
                        },
                        ..Default::default()
                    },
                )
                .await
                .unwrap();

                match individually {
                    true => {
                        for tx in txs.iter() {
                            index_transaction(&setup, tx).await;
                        }
                    }
                    false => {
                        index_multiple_transactions(&setup, txs.to_vec()).await;
                    }
                }
            }
        }
        for (owner, owner_name) in [
            (payer_pubkey.clone(), "payer"),
            (receiver_pubkey.clone(), "receiver"),
        ] {
            let accounts = setup
                .api
                .get_compressed_accounts_by_owner(GetCompressedAccountsByOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();
            assert_json_snapshot!(
                format!("{}-{}-accounts", name.clone(), owner_name),
                accounts
            );
            let proofs = setup
                .api
                .get_multiple_compressed_account_proofs(HashList(
                    accounts
                        .value
                        .items
                        .iter()
                        .map(|x| x.hash.clone())
                        .collect(),
                ))
                .await
                .unwrap();
            assert_json_snapshot!(format!("{}-{}-proofs", name.clone(), owner_name), proofs);

            let signatures = setup
                .api
                .get_compression_signatures_for_owner(photon_indexer::api::method::get_compression_signatures_for_owner::GetCompressionSignaturesForOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();

            let limit = photon_indexer::api::method::utils::Limit::new(1).unwrap();
            let mut cursor = None;
            let mut paginated_signatures = Vec::new();
            loop {
                let res = setup
                    .api
                    .get_compression_signatures_for_owner(photon_indexer::api::method::get_compression_signatures_for_owner::GetCompressionSignaturesForOwnerRequest {
                        owner: owner.clone(),
                        cursor,
                        limit: Some(limit.clone()),
                    })
                    .await
                    .unwrap()
                    .value;
                paginated_signatures.extend(res.items);
                cursor = res.cursor;
                if cursor.is_none() {
                    break;
                }
            }
            assert_eq!(signatures.value.items, paginated_signatures);
            assert_json_snapshot!(
                format!("{}-{}-signatures", owner_name, name.clone()),
                signatures
            );

            let balance = setup
                .api
                .get_compressed_balance_by_owner(photon_indexer::api::method::get_compressed_balance_by_owner::GetCompressedBalanceByOwnerRequest {
                    owner,
                    ..Default::default()
                })
                .await
                .unwrap();

            assert_eq!(
                accounts
                    .value
                    .items
                    .iter()
                    .map(|x| x.lamports.0)
                    .sum::<u64>(),
                balance.value.0,
            );
        }
    }
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_index_block_metadata(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup_with_options(
        name.clone(),
        TestSetupOptions {
            network: Network::Mainnet,
            db_backend,
        },
    )
    .await;

    let slot = 254170887;
    let block = cached_fetch_block(&setup, slot).await;
    index_block(&setup.db_conn, &block).await.unwrap();
    let filter = blocks::Column::Slot.eq(block.metadata.slot);

    let block_model = blocks::Entity::find()
        .filter(filter)
        .one(setup.db_conn.as_ref())
        .await
        .unwrap()
        .unwrap();

    assert_eq!(block_model.slot, 254170887);
    assert_eq!(block_model.parent_slot, 254170886);
    assert_eq!(
        Hash::try_from(block_model.parent_blockhash)
            .unwrap()
            .to_string(),
        "9BMHTdybcGah8PWtCzu8tVFDBXmiEHZZDf3FaZ651Nf"
    );
    assert_eq!(
        Hash::try_from(block_model.blockhash).unwrap().to_string(),
        "5GG5pzTbH6KgZM54M9XWfBNmBupQuZXdQxoZRRQXHcpM"
    );
    assert_eq!(block_model.block_height, 234724352);
    assert_eq!(block_model.block_time, 1710441678);

    // Verify that we don't get an error if we try to index the same block again
    index_block(&setup.db_conn, &block).await.unwrap();
    assert_eq!(setup.api.get_indexer_slot().await.unwrap().0, slot);

    // Verify that get_indexer_slot() gets updated a new block is indexed.
    let block = cached_fetch_block(&setup, slot + 1).await;
    index_block(&setup.db_conn, &block).await.unwrap();
    assert_eq!(setup.api.get_indexer_slot().await.unwrap().0, slot + 1);
}
