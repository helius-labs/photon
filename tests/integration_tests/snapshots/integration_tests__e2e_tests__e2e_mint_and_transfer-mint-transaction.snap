---
source: tests/integration_tests/e2e_tests.rs
expression: parsed_transaction
---
{
  "transaction": {
    "slot": 35,
    "transaction": [
      "Av+WiRe57vFwPUQKzRQHnPaUJqKbbrjPpUkF27RcYAxtPZhdq6ikoh0NxKCCGVniVMRtxtgeB7uCW+6pbdzl7A2rK0aNgQI+ioTDJsflRmqQj/vKpDkW2vgdzC9H0YwRO2Hi1jHEHr4/awlT1tg/F1o5oChVrvKm8V7ssaSP0MgNgAIABw/J2lX3bEKT6DH4ni9RzC212IxZeKmU5a+0OtHeQ+TsC7sdjF5JtM0FsTwSts1hsN5ZK+Hy76g3zEDLu4Epl/QZendTG1jI9lYwItes4XavUyXldhzjZ7KkYokhdhhhziTgdtuX3cdKxS9Ldr+osOrYw+dgHuqqbXOmHdsDJaFSeju2Dxy4AUVQTHFkwBP+TayTfjLZrAn9QVwg1Qcy/VSJDpNDIZ551u6iXE8btewjsfOdab9iDneAZd45Q3AiMOvX+tyc6TNGVUQ5YZdj5K3xeuq9mxhHkZBRPQgxmJRMP0RNfSBMgD20Ac9FLHl2mRGzt3Mio39m1gFXr7FfMUFFAwZGb+UhFzL/7K26csOb57yM5bvF9xJrLEObOkAAAACD2/n23cQhA3IXeesS5UeYJ1ep0I9lK4D1OxaGtud0IQbd9uHXZaGT2cvhRs7reawctIXtX1s3kTqM9YV+/wCpUWwytQBJW8Xd12pFBWuS/CX8e68+yKjmbwbZR2y6uFMLvA/Au0fKL3TEES6UqxPPo8Y05dwX6ssDzRojzX54fEFsPbA0deqFxq9Dqwwvj74oVYWL+D/gZzHfQIpcGaAdAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACNg6WcawbxhtlWOtrwlzf8gaLn+uNZoHQa3C681odagIIAAUCQEIPAAkOAAECAwQKCwUMBg0HCQ458SIwuiWze8ABAAAAVLI8fMGaUxepxMtX2GY+Jryt9I8aMSnMJis1MDd3Yc0BAAAA6AMAAAAAAAD+AA==",
      "base64"
    ],
    "meta": {
      "err": null,
      "status": {
        "Ok": null
      },
      "fee": 10000,
      "preBalances": [
        995022520,
        0,
        1461600,
        1461600,
        2039280,
        1169280,
        0,
        9496725120,
        1,
        1141440,
        929020800,
        1141440,
        1141440,
        1141440,
        1
      ],
      "postBalances": [
        995012370,
        0,
        1461600,
        1461600,
        2039280,
        1169280,
        0,
        9496725270,
        1,
        1141440,
        929020800,
        1141440,
        1141440,
        1141440,
        1
      ],
      "innerInstructions": [
        {
          "index": 1,
          "instructions": [
            {
              "programIdIndex": 10,
              "accounts": [
                3,
                4,
                2
              ],
              "data": "6qYT3cRnoTxb",
              "stackHeight": 2
            },
            {
              "programIdIndex": 11,
              "accounts": [
                0,
                2,
                5,
                12,
                6,
                13,
                9,
                11,
                11,
                14,
                11,
                7
              ],
              "data": "pYB2iCLnCEjZ1dTtcwRejJ19F7CQinYV6V57qKwBzgAoe2FMX8cPudKVQMHLnmBefSJopnyz4xLi4jMm7CBetvAo5N4V7oXT6c2YsPNmTpbhCt9NSwKCSAXAEQ1rEHBxVxYgdPUkjQ9U9bGHNEW2ea7tfVMpeAFVmDXicJujrD1zU2hzajYXEHHXbgyNyBtip2ArUnJQq8QkyjMG5NPvNJEnvMCdr5uA9H4PvptPqAdzZJbc612xjUw4USfboD27c6HmD7MTAyckcX8k2pihep7sxjAPQi6V4RziZqDWGraaYA7GUfJggw7sRLYHUjT3FySQwAfG98tAEKAgoBF62Wre4gowVcYBCWsXU67fuUgmWmWuyLX5jdmi1PWyL5QoCkAnQJAj4VorioX7aCk1kgh6mGNmYfRGzquHB5q",
              "stackHeight": 2
            },
            {
              "programIdIndex": 13,
              "accounts": [
                0,
                6,
                5,
                12,
                14,
                7
              ],
              "data": "5wPdbtbk7M4b3JSHrw1PY5Cxrrw1oY52hQABaC4MRQtW7r6dfYFJiPhmBDWxEQ",
              "stackHeight": 3
            },
            {
              "programIdIndex": 12,
              "accounts": [],
              "data": "LcYsUQdeQKxGUiMMFfD4pZgkFNr3k3cRuoFyrnCmq4CHUtoZLo1JaquAUtFpmShzRr9UC1ZHrwX6hpjpwPioiFAWENQwLp13uM9NH3yVrwtuyr9gjrCotggHnjEziKv3Yq5WwvPfL56Npf8ifuJPLL65RxjsbXqsfZYhhaBfHDFssAWJsCmKBbRkP3LTBcHeTexBKqSVsPYGnesoYBF7MHkuBLE3BTz9kohrG1ta1Vt47vnfPsWPS2DUMDmsSCXUPoHeRrWPe8wuaLzoffmHRD8cZYw7k7dPpYmE8bSbrjvnWQvNbxyN2qY9Wjf4KWZv2SmQ6bjH1WEpMW5ZNAsJVcgciMo4K4BNN7sKpFH8sstBvHvT4ggfb6CaS2qiUpHzrEczSHHpGGbSK3AEnJAEchcv6MK69HmXZRerhmuFAs49ZjiMhTPMy2bQJnGojR5ZD8czkujTyzEbEw1ucCgHS5pxcGrL5yQsk5TnCT2Y6AvZVi3GG7JzpcLwZc58JCoVM4jHmuJWXeSj9iQK4GJA9EqdAu9j79m81a5qXq66gefLaDLqvxZ9EiWQ4WKNqZ4UTFXEXDbW5jMAEJgtPxVCoJSqrUSbwWGTKzUSHEXz9J9AtRwmUHrxUFR4kxfeRGdzYBhLzQDuqBPBmhjE4NpFvyWgterMSAGP1Gn9dpeXDWu4orbVRtLhR3VaKCahjB2dy7FKjJ9813dqpR4xv3qRo1vtLgBoWVtqQk3o3N84fUv3L4XsGrrVs6Axj4sZMajv7eXCyZxpeAEnTBFPfU1NgpTE16N4hsuTQm8a3eSn3EmKdroJartnELUgHFJBffjq1b8vehi2c9fG4XmhJoDJ4XXc918GSuMayZBmzSdavwJwhmixy87kZvZ16dZ1d9qhL8A8SPJRJaQ37uPAACDFfT7L1hDsa3FVmh1Dvf8e8zUTThNBTqTeLGhqbZXnyeADNw6FnqMQLKj6r44d8FeVa5xN3sNXcA5fwzQHeBHybWzr741m2qdhmwHPLCwPXuUcz6G2fymtLwqXvk2bWXrgJyRzDW7JDctZmXFCTwNfPojyxDnW7VCHBS5yKDDQJPCEDvKgWoQrPYbBAYZJCG6nmWf9b2YBTCMm8unuK2kxsqRAbnfdpR8gynbDRrwxTpbyxhD2cMzESzEtxLVMr3ikGnEUJBFMbuVVdBr2Z2WrPdELQy2QKZbn4rvbvXYhorLbwEemdrXcGgLJywDQhyyb6XC8JH3RzADN97aPZYqH1NcxsuUazYu38TiXSSy3DhyXCH41Xjgs1ktp27SZWLgBQvHTeBjWc4UU9HxyRjP1XbS7UyqW5TvzrvJmsrNALWznXhGAB91WGs9",
              "stackHeight": 4
            },
            {
              "programIdIndex": 14,
              "accounts": [
                0,
                7
              ],
              "data": "3Bxs4Rwr9ZeRDby9",
              "stackHeight": 4
            },
            {
              "programIdIndex": 12,
              "accounts": [],
              "data": "11116tjoXrS9QRrV4wv8Z74xevHvd9kVED6LCnXm7vYauYxXSEwu15f2G8qm5GJ62uUFvmHiLMo8TiHG5W8SzQPLWmPFsADELBuTyLMFQGS6gBCUMM58uz3cmYV4SbWEz9NWhYiQcScvE3XFABXGZZzD4qFLaVZtDcYvuoC7tyKC86ok2ycU3uieboD21sJNgSS3fZj8qLX4QuXitWPjkjUWT7BUDa9PPjw4gRfoouGqCwN1WenWMbXdjPDeZgKzs7A6yP45jFJr7nN3ohpEKonH1YaPwRZtMGw2gvS1DdkmRGZ9C4RrvnJPKS6sGNCiduwmPis2uWCxiAyev9bDibUPqhSu3kC9rucS59Tsd",
              "stackHeight": 3
            }
          ]
        }
      ],
      "logMessages": [
        "Program ComputeBudget111111111111111111111111111111 invoke [1]",
        "Program ComputeBudget111111111111111111111111111111 success",
        "Program 9sixVEthz2kMSKfeApZXHwuboT6DZuT6crAYJTciUCqE invoke [1]",
        "Program log: Instruction: MintTo",
        "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [2]",
        "Program log: Instruction: MintTo",
        "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 4492 of 983985 compute units",
        "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success",
        "Program 6UqiSPd2mRCTTwkzhcs1M6DGYsqHWd5jiPueX3LwDMXQ invoke [2]",
        "Program log: Instruction: InvokeCpi",
        "Program 5QPEJ5zDsVou9FQS3KCauKswM3VwBEBu4dpL9xTqkWwN invoke [3]",
        "Program log: Instruction: AppendLeavesToMerkleTrees",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV invoke [4]",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV consumed 38 of 889636 compute units",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV success",
        "Program 11111111111111111111111111111111 invoke [4]",
        "Program 11111111111111111111111111111111 success",
        "Program 5QPEJ5zDsVou9FQS3KCauKswM3VwBEBu4dpL9xTqkWwN consumed 55076 of 936930 compute units",
        "Program 5QPEJ5zDsVou9FQS3KCauKswM3VwBEBu4dpL9xTqkWwN success",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV invoke [3]",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV consumed 38 of 879305 compute units",
        "Program noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV success",
        "Program 6UqiSPd2mRCTTwkzhcs1M6DGYsqHWd5jiPueX3LwDMXQ consumed 85987 of 964681 compute units",
        "Program 6UqiSPd2mRCTTwkzhcs1M6DGYsqHWd5jiPueX3LwDMXQ success",
        "Program 9sixVEthz2kMSKfeApZXHwuboT6DZuT6crAYJTciUCqE consumed 121935 of 999850 compute units",
        "Program 9sixVEthz2kMSKfeApZXHwuboT6DZuT6crAYJTciUCqE success"
      ],
      "preTokenBalances": [
        {
          "accountIndex": 4,
          "mint": "G7DSgLMiDVvMGdY9nj9gXA9XW9E1xeGhVNgCBjs6zJ5w",
          "uiTokenAmount": {
            "uiAmount": null,
            "decimals": 2,
            "amount": "0",
            "uiAmountString": "0"
          },
          "owner": "z88EeQ81Khfq5CHpSq347RVEuZERqGWW7MkTjAdxdrM",
          "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
        }
      ],
      "postTokenBalances": [
        {
          "accountIndex": 4,
          "mint": "G7DSgLMiDVvMGdY9nj9gXA9XW9E1xeGhVNgCBjs6zJ5w",
          "uiTokenAmount": {
            "uiAmount": 10.0,
            "decimals": 2,
            "amount": "1000",
            "uiAmountString": "10"
          },
          "owner": "z88EeQ81Khfq5CHpSq347RVEuZERqGWW7MkTjAdxdrM",
          "programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
        }
      ],
      "rewards": [],
      "loadedAddresses": {
        "writable": [],
        "readonly": []
      },
      "computeUnitsConsumed": 122085
    },
    "version": 0,
    "blockTime": 1715117976
  },
  "compression_info": {
    "closed_accounts": [],
    "opened_accounts": [
      {
        "account": {
          "hash": "3i8NnHAxhT5oFiuCJyEnHTZCUNMo1SnURMkqk4mnJsY8",
          "address": null,
          "data": {
            "discriminator": 2,
            "data": "4Hbbl93HSsUvS3a/qLDq2MPnYB7qqm1zph3bAyWhUnpUsjx8wZpTF6nEy1fYZj4mvK30jxoxKcwmKzUwN3dhzegDAAAAAAAAAAAAAAAAAAAAAAA=",
            "dataHash": "b5c2fTX2WAfSuwD3zUDWuSCDpVhRr299gDJqiyBVzx9"
          },
          "owner": "9sixVEthz2kMSKfeApZXHwuboT6DZuT6crAYJTciUCqE",
          "lamports": 0,
          "tree": "5bdFnXU47QjzGpzHfXnxcEi5WXyxzEAZzd1vrE39bf1W",
          "leafIndex": 0,
          "seq": 1,
          "slotCreated": 35
        },
        "optional_token_data": {
          "mint": "G7DSgLMiDVvMGdY9nj9gXA9XW9E1xeGhVNgCBjs6zJ5w",
          "owner": "6hct2FBjZeQZ7Sscip1Xep3b94jCh7FvtaKqpfhne1Ak",
          "amount": 1000,
          "delegate": null,
          "state": "initialized",
          "isNative": null,
          "delegatedAmount": 0
        }
      }
    ]
  }
}
